trie/iterator.go

```diff
@@ -52,74 +59,113 @@ func (it *Iterator) Next() bool {
 	return false
 }
 
-func (it *Iterator) makeKey() []byte {
-	key := it.keyBuf[:0]
-	for _, se := range it.nodeIt.stack {
-		switch node := se.node.(type) {
-		case *fullNode:
-			if se.child <= 16 {
-				key = append(key, byte(se.child))
-			}
-		case *shortNode:
-			if hasTerm(node.Key) {
-				key = append(key, node.Key[:len(node.Key)-1]...)
-			} else {
-				key = append(key, node.Key...)
-			}
-		}
-	}
-	return decodeCompact(key)
+// NodeIterator is an iterator to traverse the trie pre-order.
+type NodeIterator interface {
```

Comment:
A few special mentions probably warranted for:

 * The caller of the interface methods must not modify the returned values (at least for LeadBlob and Path), since those are not copied by the iterator, just returned as is.
 * The caller should not retain the result of Path or should make a copy, since the next invocation to "Next" will potentially change its content.